src/useAddonState/manager.ts

```diff
@@ -0,0 +1,54 @@
+import type { API } from "@storybook/manager-api";
+import { useStorybookApi } from "@storybook/manager-api";
+import { useCallback } from "react";
+
+import { getValue, GetValuePayload, setValue, SetValuePayload } from "./common";
+
+const requestedKeys = new Set<string>();
+let listening = false;
+function ensureListening(api: API) {
+  if (listening) return;
+
+  api.on(getValue, ({ key }: GetValuePayload) => {
+    const apiKey = `experimental_${key}`;
+    const value = api.getAddonState(apiKey);
+    if (value !== undefined) api.emit(setValue, { key, value } satisfies SetValuePayload);
+  });
+
+  api.on(setValue, ({ key, value }: SetValuePayload) => {
+    const apiKey = `experimental_${key}`;
+    api.setAddonState(apiKey, value);
+  });
+
+  listening = true;
+}
+
+function apiGetAddonState<T>(api: API, key: string) {
+  const apiKey = `experimental_${key}`;
+
+  // Ask the universe for the value of the key if we've never seen it before
+  if (!requestedKeys.has(key)) {
+    api.emit(getValue, { key } satisfies GetValuePayload);
+    requestedKeys.add(key);
+  }
+
+  return api.getAddonState(apiKey) as T | undefined;
```

Comment:
I imagine that `emit` won't immediately/synchronously set the value?
So, `api.getAddonState(apiKey)` will return undefined?

Maybe `apiGetAddonState` should return a `Promise`?

Reply:
The api will return `undefined` until the response to the get call comes back from somewhere else. 

We could make it a promise in node, but we don't actually know if any other place has a value for the key so it might never resolve.

Reply:
I agree with @ndelangen, better to have an API that returns a Promise, than one that (synchronously) says "no" but then says "yes" when called again. A timeout can avoid the never-resolving issue.

Reply:
But you can't do anything with a promise in a React component, can. you? And for consistency would you always have to return a promise?

We could look at using suspense or something but that seem complicated to me. This feels more idiomatic for a React hook.